### PR TITLE
EXTENSION: Feature/dx 256 ongoing auction claiming and linking

### PR DESCRIPTION
**Pull request by W3stside**
# Extends #240 

1. Added `Link` component to move user to `AuctionStatus` page related to `ongoingAuction` clicked
2. Created `claimSellerFundsFromSeveral` thunk/ `api/index` and `api/dutchx` to enable claiming from auctions with sellerbalances and clear status
3. removed state print in `AuctionStatusHOC`

### CODE REVIEW

File: src/components/MenuAuctions/index.tsx [hunk 1]
```diff
@@ -26,13 +30,17 @@ export const MenuAuctions: React.SFC<MenuAuctionProps> = ({
               {ongoingAuctions.map(
                 (auction, i) =>
                   <tr key={`${auction.sell.address}-${auction.buy.address}-${i}`}>
-                    <td>{`${auction.sell.symbol}/${auction.buy.symbol}`}</td>
+                    <td>
+                      <Link to={`/auction/${auction.sell.symbol}-${auction.buy.symbol}-${auction.indices[auction.indices.length - 1]}`}>
```
Comment (Velenir) [inline src/components/MenuAuctions/index.tsx]: This `Link` is always to the latest (ongoing) auction, what if user doesn't have anything commited in that auction, only claimable in past auctions? Where should the link lead?

File: src/api/index.ts [hunk 2]
```diff
@@ -533,31 +562,35 @@ export const getSellerOngoingAuctions = async (
...
 
       const { sell: { decimals }, buy: { decimals: decimalsInverse } } = auction
-      const { lastIndex, closingPriceDir, closingPriceOpp } = lastAuctionsData[index]
-
-      const ongoingAuction = {
-        ...auction,
-        indices,
-        indicesInverse,
+      if (!(indices.length >= 1 || indicesInverse.length >= 1)) {
+        return accum
```
Comment (Velenir) [inline src/api/index.ts]: why the change? it does literally the same, but creates `ongoingAuction` with let, unnecessarily, has  potentially skippable on the first conditional `auction` destructuring, unnecessarily, uses `else` instead of an early `return`

File: src/api/index.ts [hunk 3]
```diff
@@ -363,6 +363,34 @@ export const getUnclaimedSellerFunds = async (pair: TokenPair, index?: Index, ac
...
+  buy: DefaultTokenObject,
+  userAccount?: Account,
+  indices: number = 0,
+) => {
+  const { DutchX } = await promisedAPI
+  userAccount = await fillDefaultAccount(userAccount)
+  const sellArr = [], buyArr = []
+
+  const claimableIndices = (await DutchX.getIndicesWithClaimableTokensForSellers(sell.address, buy.address, userAccount, indices))[0].map((i: BigNumber) => i.toNumber())
+  const claimableIndicesBalance = (await DutchX.getIndicesWithClaimableTokensForSellers(sell.address, buy.address, userAccount, indices))[1].map((i: BigNumber) => i.toNumber())
```
Comment (Velenir) [inline src/api/index.ts]: This and above can be inside `Promise.all`. Also is `claimableIndicesBalance` only used for logging?

File: src/api/dutchx.ts [hunk 4]
```diff
@@ -92,19 +92,32 @@ async function init(): Promise<DutchExchange> {
...
+    account: Account,
+  ) => dx.claimTokensFromSeveralAuctionsAsSeller(
+    sellTokenAddresses,
+    buyTokenAddresses,
+    indices,
+    account,
+    { from: account },
+  )
+
+  const claimBuyerFunds = (
```
Comment (Velenir) [inline src/api/dutchx.ts:110]: Do we need to use `claimBuyerFunds` anywhere or is it here for completion's sake?

File: src/actions/blockchain.ts [hunk 5]
```diff
@@ -341,6 +343,32 @@ export const approveAndPostSellOrder = (choice: string) => async (dispatch: Func
   }
 }
 
+export const claimSellerFundsFromSeveral = (
+  sell: DefaultTokenObject,
+  buy: DefaultTokenObject,
+  indices?: number,
```
Comment (Velenir) [inline src/actions/blockchain.ts]: Can you rename `indices` to something like `lastNIndex`. I was expecting it to be an array based on the name...

File: src/api/index.ts [hunk 6]
```diff
@@ -363,6 +363,34 @@ export const getUnclaimedSellerFunds = async (pair: TokenPair, index?: Index, ac
   }
 }
 
+export const getUnclaimedSellerFundsFromSeveral = async (
```
Comment (Velenir) [inline src/api/index.ts]: This function actually claims the tokens, not just shows how much is claimable? The `get` in the name confuses me.

File: src/api/index.ts [hunk 7]
```diff
@@ -363,6 +363,34 @@ export const getUnclaimedSellerFunds = async (pair: TokenPair, index?: Index, ac
...
+  sell: DefaultTokenObject,
+  buy: DefaultTokenObject,
+  userAccount?: Account,
+  indices: number = 0,
+) => {
+  const { DutchX } = await promisedAPI
+  userAccount = await fillDefaultAccount(userAccount)
+  const sellArr = [], buyArr = []
+
+  const claimableIndices = (await DutchX.getIndicesWithClaimableTokensForSellers(sell.address, buy.address, userAccount, indices))[0].map((i: BigNumber) => i.toNumber())
```
Comment (Velenir) [inline src/api/index.ts]: I know it shouldn't happen based on the place the function is called, but please add a check like:
```
if (claimableIndices .length === 0) return
```
or some such.

File: src/api/index.ts [hunk 8]
```diff
@@ -363,6 +363,34 @@ export const getUnclaimedSellerFunds = async (pair: TokenPair, index?: Index, ac
...
+  const claimableIndicesBalance = (await DutchX.getIndicesWithClaimableTokensForSellers(sell.address, buy.address, userAccount, indices))[1].map((i: BigNumber) => i.toNumber())
+
+  const lastIndexCleared = ((await DutchX.getClosingPrice({ sell, buy }, claimableIndices[claimableIndices.length - 1])).map(i => i.toNumber()))[0] > 0
```
Comment (Velenir) [inline src/api/index.ts]: Can be one action shorter with
```
(await DutchX.getClosingPrice({ sell, buy }, claimableIndices[claimableIndices.length - 1])[0].greater(0)
```
no need to `.map`

File: src/api/index.ts [hunk 9]
```diff
@@ -363,6 +363,34 @@ export const getUnclaimedSellerFunds = async (pair: TokenPair, index?: Index, ac
...
+
+  console.log('claimableIndices => ', claimableIndices, 'claimableIndicesBalance => ', claimableIndicesBalance, 'lastIndexCleared? => ', lastIndexCleared)
+
+  for (let i = 1; i <= claimableIndices.length; ++i) {
+    if (!lastIndexCleared) continue
```
Comment (Velenir) [inline src/api/index.ts]: You want to include `sell/buy.address` for all indices except for the last if it happens to be the ongoing auction not yet closed, right? `if (!lastIndexCleared) continue` will end up in empty arrays should it trigger.

What about:
```
cost length = lastIndexCleared ? claimableIndices.length : claimableIndices.length - 1
const sellArr = Array.from({ length }, () => sell.address)
const buyArr = ...
```
This will give for
```
claimableIndices [1,3,5] and lastIndexCleared = true  =>  sellArr = [address, address, address]

claimableIndices [1,3,5] and lastIndexCleared = false =>  sellArr = [address, address]
```

**Review by Velenir — CHANGES_REQUESTED** [review]
Good stuff, much awaited claiming functionality

File: src/components/MenuAuctions/index.tsx (hunk 1 above)
Comment (W3stside) [reply src/components/MenuAuctions/index.tsx]: maybe latest with claiming if current has no committed?

File: src/api/index.ts (hunk 2 above)
Comment (W3stside) [reply src/api/index.ts]: artifact from tests I am running for showing the first auction in `ongoingAuctions` if auction hasn't started yet.

consider the current UX now - on rinkeby or anywhere else - where a user commits funds to auction 1 that hasn't started. `ongoingAuctions` shows nothing, which is unclear and quite frankly, bad UX. We need to show something for the first auction index. So I changed the code here for just easier testing in order to at the very least return the auction pair returned from `getRunningPairs` and show `committed` amount

File: src/api/index.ts (hunk 6 above)
Comment (W3stside) [reply src/api/index.ts]: ok will change

File: src/api/index.ts (hunk 7 above)
Comment (W3stside) [reply src/api/index.ts]: will do

File: src/api/index.ts (hunk 8 above)
Comment (W3stside) [reply src/api/index.ts]: true

File: src/api/index.ts (hunk 9 above)
Comment (W3stside) [reply src/api/index.ts]: cool, nice optimization. will change

File: src/api/index.ts (hunk 3 above)
Comment (W3stside) [reply src/api/index.ts]: yes, artifact

File: src/api/dutchx.ts (hunk 4 above)
Comment (W3stside) [reply src/api/dutchx.ts:110]: as of now, completion

File: src/actions/blockchain.ts (hunk 5 above)
Comment (W3stside) [reply src/actions/blockchain.ts]: yes, also an artifact (lots of those from weekend coding, lol) as before I did not check using `getIndicesWithClaimableTokensForSellers`

File: src/api/index.ts (hunk 6 above)
Comment (W3stside) [reply src/api/index.ts]: it claims. it's also the same name as in `DX.sol`

**Comment by W3stside** [comment]
weeeekkkeennnddd coooodddiiinnngggg, this is the most you've wrecked my PR 😂 😂 😂 

all good stuff, thanks for the thorough checkthrough

File: src/components/MenuAuctions/index.tsx (hunk 1 above)
Comment (W3stside) [reply src/components/MenuAuctions/index.tsx]: @Velenir are we ok with what I suggested?

File: src/actions/blockchain.ts (hunk 5 above)
Comment (W3stside) [reply src/actions/blockchain.ts]: done

File: src/components/MenuAuctions/index.tsx (hunk 1 above)
Comment (Velenir) [reply src/components/MenuAuctions/index.tsx]: Yes, sounds reasonable

File: src/actions/blockchain.ts [hunk 10]
```diff
@@ -359,8 +359,11 @@ export const claimSellerFundsFromSeveral = (
         body: `Claiming ${buyName} tokens from ${sellName}-${buyName} auction. Please check ${activeProvider}`,
       },
     }))
-    const claimReceipt = await getUnclaimedSellerFundsFromSeveral(sell, buy, currentAccount, indices)
-    console.log('​claimReceipt => ', claimReceipt)
+    const [claimReceipt] = await Promise.all([
+      getUnclaimedSellerFundsFromSeveral(sell, buy, currentAccount, indices),
+      updateMainAppState(),
```
Comment (Velenir) [inline src/actions/blockchain.ts]: `updateMainState` should be fired after `getUnclaimedSellerFundsFromSeveral` tx is executed and mined, otherwise it can fetch the old state. And even then it won't necessarily bring in the new state as happened in owl project (MMask is to blame?)
You don't need to `await` on `updateMainState`, you can just start it after receipt is received and it'll run it's course by itself without stalling `claimSellerFundsFromSeveral`  function execution

**Review by Velenir — CHANGES_REQUESTED** [review]

File: src/actions/blockchain.ts (hunk 10 above)
Comment (W3stside) [reply src/actions/blockchain.ts]: good call - also yes, I checked this in the app and it doesn't update appropriately

**Review by Velenir — APPROVED** [review]
